components/script/dom/macros.rs

```diff
@@ -632,3 +632,132 @@ macro_rules! handle_potential_webgl_error {
         handle_potential_webgl_error!($context, $call, ());
     };
 }
+
+macro_rules! impl_document_or_shadow_root_helpers(
+    () => (
+        /// <https://html.spec.whatwg.org/multipage/#concept-document-bc>
+        #[inline]
+        pub fn browsing_context(&self) -> Option<DomRoot<WindowProxy>> {
+            if self.has_browsing_context {
+                self.window.undiscarded_window_proxy()
+            } else {
+                None
+            }
+        }
+
+        pub fn nodes_from_point(
+            &self,
+            client_point: &Point2D<f32>,
+            reflow_goal: NodesFromPointQueryType,
+        ) -> Vec<UntrustedNodeAddress> {
+            if !self
+                .window
+                .layout_reflow(QueryMsg::NodesFromPointQuery(*client_point, reflow_goal))
+            {
+                return vec![];
+            };
+
+            self.window.layout().nodes_from_point_response()
+        }
+    );
+);
+
+macro_rules! impl_document_or_shadow_root_methods(
+    ($struct:ident) => (
+        #[allow(unsafe_code)]
+        // https://drafts.csswg.org/cssom-view/#dom-document-elementfrompoint
+        fn ElementFromPoint(&self, x: Finite<f64>, y: Finite<f64>) -> Option<DomRoot<Element>> {
+            let x = *x as f32;
+            let y = *y as f32;
+            let point = &Point2D::new(x, y);
+            let viewport = self.window.window_size().initial_viewport;
+
+            if self.browsing_context().is_none() {
+                return None;
+            }
+
+            if x < 0.0 || y < 0.0 || x > viewport.width || y > viewport.height {
+                return None;
+            }
+
+            match self
+                .nodes_from_point(point, NodesFromPointQueryType::Topmost)
+                .first()
+            {
+                Some(address) => {
+                    let js_runtime = unsafe { JS_GetRuntime(self.window.get_cx()) };
+                    let node = unsafe { node::from_untrusted_node_address(js_runtime, *address) };
+                    let parent_node = node.GetParentNode().unwrap();
+                    let element_ref = node
+                        .downcast::<Element>()
+                        .unwrap_or_else(|| parent_node.downcast::<Element>().unwrap());
+
+                    Some(DomRoot::from_ref(element_ref))
+                },
+                None => self.GetDocumentElement(),
+            }
+        }
+
+        #[allow(unsafe_code)]
+        // https://drafts.csswg.org/cssom-view/#dom-document-elementsfrompoint
+        fn ElementsFromPoint(&self, x: Finite<f64>, y: Finite<f64>) -> Vec<DomRoot<Element>> {
+            let x = *x as f32;
+            let y = *y as f32;
+            let point = &Point2D::new(x, y);
+            let viewport = self.window.window_size().initial_viewport;
```

Comment:
This needs to retarget nodes in shadow trees and such... Rather than having these implementations duplicated I would have them just in DocumentOrShadowRoot, which also avoids having to implement dummy ShadowRoot::GetBody & co.